### PR TITLE
Feat: LSP flow works

### PR DESCRIPTION
**Pull request by Tulun**
**Motivation**
UI should expire and settle the contract and allow users to redeem remaining balances of their long and short tokens. 

**Details**
Note: Testing the flow is rather complicated, involving using a bunch of scripts in hardhat-test, setting up MM with your hardhat instance, and using hte UI.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [x]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

**Comment by mrice32** [comment]
@Tulun can you tag @Gamaranto on this PR? All FE PRs should be reviewed by somebody on the @UMAprotocol/frontend team.

File: pages/testing.tsx [hunk 1]
```diff
@@ -47,11 +53,59 @@ const Testing = () => {
...
+    ) {
+      setShowSettle(true);
+    }
+  }, [contractExpirationTime, currentTime, contractState]);
   // Get contract data and set values.
   useEffect(() => {
     if (web3Provider && !lspContract) {
       const signer = web3Provider.getSigner();
       const contract = createLSPContractInstance(signer, KNOWN_LSP_ADDRESS);
+      contract.contractState().then(async (cs: ContractState) => {
```
Comment (mrice32) [inline pages/testing.tsx:79]: Should we have this useeffect poll at all? Right now, it looks like it only runs on network change or startup.

**Review by mrice32 — APPROVED** [review]
Looks great! One minor question

File: pages/testing.tsx (hunk 1 above)
Comment (Tulun) [reply pages/testing.tsx:79]: Yeah I think so. Keep in mind the web3provider will be defined *after* the user connects their wallet above -- or at least some of these values.

So there will be some queries we will make that might change once this view integrates with Onboard.

File: components/lsp-form/LSPForm.tsx [hunk 2]
```diff
@@ -42,8 +47,48 @@ const LSPForm: FC<Props> = ({
...
+  setShowSettle,
+  contractState,
+  setContractState,
 }) => {
-  const [showSettle, setShowSettle] = useState(false);
+  const expire = useCallback(async () => {
+    if (lspContract) {
+      try {
+        await lspContract.expire().then((tx: any) => {
+          return tx.wait(1).then(() => {
```
Comment (daywiss) [inline components/lsp-form/LSPForm.tsx:59]: nit, flattening the promise chain is just a bit more clean generally ie:
```js
await lspContract.expire()
  .then((tx: any) => tx.wait(1))
  .then(() => {
   // ...etc
````

File: components/lsp-form/LSPForm.tsx [hunk 3]
```diff
@@ -42,8 +47,48 @@ const LSPForm: FC<Props> = ({
...
+    }
+  }, [lspContract, setShowSettle, setContractState]);
+
+  const settle = useCallback(async () => {
+    if (lspContract) {
+      try {
+        await lspContract
+          .settle(longTokenBalance, shortTokenBalance)
+          .then((tx: any) => {
+            return tx.wait(1).then(async () => {
```
Comment (daywiss) [inline components/lsp-form/LSPForm.tsx:76]: this is also a candidate to flatten

File: components/lsp-form/LSPForm.tsx [hunk 4]
```diff
@@ -87,12 +132,48 @@ const LSPForm: FC<Props> = ({
...
+              <SettleTokenBalance>
+                Long Token Balance:{" "}
+                {ethers.utils.formatUnits(longTokenBalance, collateralDecimals)}
+              </SettleTokenBalance>
+              <SettleTokenBalance>
+                Short Token Balance:{" "}
+                {ethers.utils.formatUnits(
+                  shortTokenBalance,
+                  collateralDecimals
+                )}
```
Comment (daywiss) [inline components/lsp-form/LSPForm.tsx:162]: we might want a way to show the user how much collateral they will be recieving based on price @Tommy1231232

**Review by daywiss — APPROVED** [review]
looks good. i tried testing it with hard hat scripts, got almost to end, but proposal was failing in hardhat script, not a problem in this code.

**Pull request merged**